### ISSUE TITLE
meta: Task Builder stabilization

### ISSUE THREAD

**Issue by seanmonstar** [tokio-rs/tokio#4114]
This is to track the currently unstable `tokio::task::Builder` API. To use it, you must pass `--cfg tokio_unstable` to `rustc`.

RFC: https://hackmd.io/Vm0UXvrNTg6HleGlD8C8sg
PR: https://github.com/tokio-rs/tokio/pull/3881

## Roadmap

- [x] Include code from PR in a crates.io release
- [ ] Resolve open questions

## Open questions

- How do we allow adding more context to the spawned task's `tracing` span?
  - If `tracing` is at 1.0, we could allow passing `Builder::span` directly.
  - To hide `tracing`, we could use some imaginary `Builder::tag(name, value)` that uses `Span::record` internally.
  - Or we could maybe make use of [valuable](https://tokio.rs/blog/2021-05-valuable)

**Comment by dan-da** [comment]
I might be missing something, but I don't see yet how to access the `name` after a task is spawned with `Builder`.

When using [std::thread::Builder](https://doc.rust-lang.org/std/thread/struct.Builder.html), one can access the name via [join_handle.thread().name()](https://doc.rust-lang.org/std/thread/struct.Thread.html#method.name), but I don't see any equivalent in tokio's [JoinHandle](https://docs.rs/tokio/latest/tokio/task/struct.JoinHandle.html).

use case: This would be useful for me in order to log task name when acquiring a `lock`, for example.

**Comment by Darksonn** [comment]
Currently, the name is only accessible via the tracing traces that Tokio emits when you enable the unstable tracing feature.

**Comment by dan-da** [comment]
> Currently, the name is only accessible via the tracing traces that Tokio emits when you enable the unstable tracing feature.

yes, the name is dropped [here](https://github.com/tokio-rs/tokio/blob/d51308b0359a3a3f4d3c11fd71bb402646afe0b5/tokio/src/runtime/blocking/pool.rs#L382) and I find that unfortunate.

I don't see why `Builder` and `name` should be gated behind `tracing` feature, when the concept of a named task seems generally usefully, not only for tracing.   And again, it would be symmetric with functionality already offered in (stable) std::thread.

The alternative for an app wanting to keep track of named tasks is to do it themselves at each call to spawn(), eg with a struct that contains a name and a JoinHandle.  This is a viable workaround, but is much less convenient eg to access within the spawned thread, and if I were moving code that already uses `std::thread::Builder` to tokio tasks, it would be burdensome.

Maybe this should be opened as a new issue?    Anyway it seems (to me) like something that should be addressed before the Builder feature is stabilized.

**Comment by Darksonn** [comment]
The reason it's gated behind the tracing feature is that the feature was added for that specific purpose.

Still, I agree that it would be reasonable to expand this and stabilize it for other use-cases.
